### PR TITLE
Validate Exposed and Global extended attributes

### PR DESCRIPTION
**Pull request by foolip**
This test will catch situations like this one:
https://github.com/w3c/webrtc-identity/pull/36

### CODE REVIEW

**Comment by foolip** [comment]
I've confirmed that this test failed before e95adca52e22ab9e57ffcee582c4423a0cca44e4.

File: test/idl/consistency.js [hunk 1]
```diff
@@ -339,6 +339,70 @@ views.forEach(({ name, folder }) => {
       }
     });
 
+    // Validate that Exposed attributes use only the identifiers from Global
+    // attributes.
+    it('Exposed and Global extended attributes', function () {
```
Comment (tidoust) [inline test/idl/consistency.js:344]:
```suggestion
    it('contains consistent Exposed and Global extended attributes', function () {
```

File: test/idl/consistency.js [hunk 2]
```diff
@@ -339,6 +339,70 @@ views.forEach(({ name, folder }) => {
...
+      }
+
+      // Collect the know global identifiers.
+      const globals = new Set();
+      for (const dfn of dfns) {
+        const ids = getExtAttrIdentifiers(dfn, 'Global');
+        if (!ids) {
+          continue;
+        }
+        assert(Array.isArray(ids));
```
Comment (tidoust) [inline test/idl/consistency.js:374]:
```suggestion
        assert(Array.isArray(ids), `Unexpected Global attribute value on ${dfn.name}: ${ids}`);
```

File: test/idl/consistency.js [hunk 3]
```diff
@@ -339,6 +339,70 @@ views.forEach(({ name, folder }) => {
...
+        for (const id of ids) {
+          globals.add(id);
+        }
+      }
+
+      // Check that some well known global were found.
+      assert(globals.has('Window'));
```
Comment (tidoust) [inline test/idl/consistency.js:381]:
```suggestion
      assert(globals.has('Window'), 'List of known globals does not include "Window"');
```

File: test/idl/consistency.js [hunk 4]
```diff
@@ -339,6 +339,70 @@ views.forEach(({ name, folder }) => {
...
+      assert(globals.has('ServiceWorker'));
```
Comment (tidoust) [inline test/idl/consistency.js:382]:
```suggestion
      assert(globals.has('ServiceWorker'), 'List of known globals does not include "ServiceWorker"');
```

File: test/idl/consistency.js [hunk 5]
```diff
@@ -339,6 +339,70 @@ views.forEach(({ name, folder }) => {
...
+
+      // Validate the Exposed extended attributes of top-level definitions and
+      // their members.
+      for (const dfn of dfns) {
+        const ids = getExtAttrIdentifiers(dfn, 'Exposed');
+        if (Array.isArray(ids)) {
+          for (const id of ids) {
+            assert(globals.has(id), ``);
```
Comment (tidoust) [inline test/idl/consistency.js:390]:
```suggestion
            assert(globals.has(id), `Exposed attribute value ${id} on ${dfn.name} does not match any known global name`);
```

File: test/idl/consistency.js [hunk 6]
```diff
@@ -339,6 +339,70 @@ views.forEach(({ name, folder }) => {
...
+          }
+        }
+        if (dfn.members) {
+          for (const member of dfn.members) {
+            const ids = getExtAttrIdentifiers(member, 'Exposed');
+            if (Array.isArray(ids)) {
+              for (const id of ids) {
+                assert(globals.has(id));
```
Comment (tidoust) [inline test/idl/consistency.js:398]:
```suggestion
                assert(globals.has(id), `Exposed attribute value ${id} on ${dfn.name}.${member.name} does not match any known global name`);
```

File: test/idl/consistency.js [hunk 7]
```diff
@@ -339,6 +339,70 @@ views.forEach(({ name, folder }) => {
...
+        switch (attr.rhs.type) {
+          case 'identifier':
+            return [attr.rhs.value];
+          case 'identifier-list':
+            return attr.rhs.value.map(({value}) => value);
+          case '*':
+            return '*';
+          default:
+            console.log(attr.rhs.type)
+            throw new Error(`Unexpected RHS for extended attribute ${name} on ${node.name}`);
```
Comment (tidoust) [inline test/idl/consistency.js:363]:
```suggestion
            throw new Error(`Unexpected RHS for extended attribute ${name} on ${node.name}: ${attr.rhs?.type}`);
```

File: test/idl/consistency.js [hunk 8]
```diff
@@ -339,6 +339,70 @@ views.forEach(({ name, folder }) => {
...
+        }
+        switch (attr.rhs.type) {
+          case 'identifier':
+            return [attr.rhs.value];
+          case 'identifier-list':
+            return attr.rhs.value.map(({value}) => value);
+          case '*':
+            return '*';
+          default:
+            console.log(attr.rhs.type)
```
Comment (tidoust) [inline test/idl/consistency.js:362]:
```suggestion
```

**Review by tidoust — CHANGES_REQUESTED** [review]
Thanks! That looks good, I just suggest a few tweaks to give a more useful failure message than "expected true, got false" when test fails.

**Comment by tidoust** [comment]
Comments being editorial, I'll integrate them and merge.

**Review by tidoust — APPROVED** [review]

**Comment by foolip** [comment]
Thanks @tidoust!

**Pull request merged**